nixos/modules/services/web-apps/dokuwiki.nix

```diff
@@ -0,0 +1,281 @@
+        Example: <link xlink:href="https://github.com/splitbrain/dokuwiki/blob/master/conf/users.auth.php.dist"/>
+        '';
+    };
+
+    extraConfig = mkOption {
+      type = types.nullOr types.lines;
+      default = null;
+      example = ''
+        $conf['title'] = 'My Wiki';
+        $conf['useacl'] = 1;
+        $conf['superuser'] = '@admin';
+        $conf['userewrite'] = 1;
+      '';
+      description = ''
+        DokuWiki configuration. Refer to
+        <link xlink:href="https://www.dokuwiki.org/config"/>
+        for details on supported values.
+      '';
+    };
+
+    poolConfig = mkOption {
+      type = with types; attrsOf (oneOf [ str int bool ]);
+      default = {
+        "pm" = "dynamic";
+        "pm.max_children" = 32;
+        "pm.start_servers" = 2;
+        "pm.min_spare_servers" = 2;
+        "pm.max_spare_servers" = 4;
+        "pm.max_requests" = 500;
+      };
+      description = ''
+        Options for the dokuwiki PHP pool. See the documentation on <literal>php-fpm.conf</literal>
+        for details on configuration directives.
+      '';
+    };
+
+    nginx = mkOption {
+      type = types.submodule (
+        recursiveUpdate
+          (import ../web-servers/nginx/vhost-options.nix { inherit config lib; })
+          {
+            # Enable encryption by default,
+            options.forceSSL.default = true;
+            options.enableACME.default = true;
+          }
+      );
+      default = {forceSSL = true; enableACME = true;};
+      example = {
+        serverAliases = [
+          "wiki.\${config.networking.domain}"
+        ];
+        enableACME = false;
+      };
+      description = ''
+          With this option, you can customize an nginx virtualHost which already has sensible defaults for DokuWiki.
+      '';
+    };
+  };
+
+  # implementation
+
+  config = mkIf cfg.enable {
+
+    warnings = mkIf (cfg.extraConfig == null) ["Not setting at least useacl and superuser in extraConfig may impair your ability to administer DokuWiki"];
+
+    assertions = [ 
+      {
+        assertion = cfg.acl != null || cfg.aclFile != null;
+        message = "Either services.dokuwiki.acl or services.dokuwiki.aclFile is mandatory";
+      }
+      {
+        assertion = cfg.users != null || cfg.usersFile != null;
+        message = "Either services.dokuwiki.users or services.dokuwiki.usersFile is mandatory";
+      }
+    ];
+
+    services.phpfpm.pools.dokuwiki = {
+      inherit user;
+      inherit group;
+      phpEnv = {
+        DOKUWIKI_ACL_AUTH_CONFIG = if (cfg.acl != null) then "${dokuwikiAclAuthConfig}" else "${cfg.aclFile}";
+        DOKUWIKI_LOCAL_CONFIG = "${dokuwikiLocalConfig}";
+        DOKUWIKI_PLUGINS_LOCAL_CONFIG = "${dokuwikiPluginsLocalConfig}";
+        DOKUWIKI_USERS_AUTH_CONFIG = if (cfg.users != null) then "${dokuwikiUsersAuthConfig}" else "${cfg.usersFile}";
+      };
+      
+      settings = {
+        "listen" = "/var/run/phpfpm/phpfpm.sock";
```

Comment:
Thanks for noticing, hopefully fixed! I was setting both in fpm and nginx settings but you're right, there's no need to do it that way. I've removed it.

Reply:
> Thanks for writing this module. Writing this was on my todo list but I hadn't got around to it yet.
> 
> I was going to write this to use `services.httpd` instead of `nginx`, though. Maybe this is a good opportunity to have a module support both thinking

Thanks for the review, I really do appreciate it! And sorry if this implementation is not up to par, I am more than happy for every suggestion. Still learning a lot and this is my first attempt on a module :). Regarding httpd - this will require further time as I don't use it regularly.